Use mach_vm_remap instead of mach_vm_map to set up permanent mapping for g_config

ddegazio opened the pull request:
#### 153af8646bb89084c5525d2b1983144a5a4178f4
<pre>
Use mach_vm_remap instead of mach_vm_map to set up permanent mapping for g_config
<a href="https://bugs.webkit.org/show_bug.cgi?id=290066">https://bugs.webkit.org/show_bug.cgi?id=290066</a>
<a href="https://rdar.apple.com/147347736">rdar://147347736</a>

Reviewed by Mark Lam.

Uses mach_vm_remap to set up permanent mapping for g_config, preventing
a possible issue where g_config values that were set up prior to WTF
initialization would be zeroed out.

* Source/WTF/wtf/WTFConfig.cpp:
(WTF::setPermissionsOfConfigPage):
(WTF::Config::initialize):

Canonical link: <a href="https://commits.webkit.org/292440@main">https://commits.webkit.org/292440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2fd6795d661c13959b49525d6faf1a63171e973f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96057 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15671 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5622 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101120 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46566 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15966 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24103 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/73226 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30452 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99060 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11955 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86769 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53563 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11700 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4521 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45901 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88730 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81848 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4623 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103147 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94678 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23124 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/82268 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23375 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82782 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81640 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20477 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26242 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/3683 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16480 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23087 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/118155 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22746 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26226 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24487 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/WTF/wtf/WTFConfig.cpp:
```diff
@@ -73,9 +73,13 @@ void setPermissionsOfConfigPage()
 
         auto attemptVMMapping = [&] {
-            return mach_vm_map(mach_task_self(), &addr, ConfigSizeToProtect, pageSize() - 1, flags, MEMORY_OBJECT_NULL, 0, false, VM_PROT_READ | VM_PROT_WRITE, VM_PROT_READ | VM_PROT_WRITE, VM_INHERIT_DEFAULT);
+            // We use remap here to avoid zeroing the config, since WebKit
```
nit: should specify "remap _instead of map_", since I think many people aren't aware that mach_vm_map/mmap will zero memory.

MenloDorian reviewed (approved):
r=me

kmiller68 commented on Source/WTF/wtf/spi/cocoa/MachVMSPI.h:
```diff
@@ -39,6 +39,8 @@ kern_return_t mach_vm_allocate(vm_map_t target, mach_vm_address_t*, mach_vm_size
 kern_return_t mach_vm_deallocate(vm_map_t target, mach_vm_address_t, mach_vm_size_t);
 kern_return_t mach_vm_map(vm_map_t targetTask, mach_vm_address_t*, mach_vm_size_t, mach_vm_offset_t mask, int flags,
     mem_entry_name_port_t, memory_object_offset_t, boolean_t copy, vm_prot_t currentProtection, vm_prot_t maximumProtection, vm_inherit_t);
+kern_return_t mach_vm_remap(vm_map_t targetTask, mach_vm_address_t*, mach_vm_size_t, mach_vm_offset_t mask, int flags,
```
Can we remove the duplicate SPI declaration from ExecutableAllocator so it's obvious that we're not adopting any new SPI?

ddegazio replied:
Done in most recent push.

ddegazio committed:
Use mach_vm_remap instead of mach_vm_map to set up permanent mapping for g_config
https://bugs.webkit.org/show_bug.cgi?id=290066
rdar://147347736

Reviewed by Mark Lam.

Uses mach_vm_remap to set up permanent mapping for g_config, preventing
a possible issue where g_config values that were set up prior to WTF
initialization would be zeroed out.

* Source/WTF/wtf/WTFConfig.cpp:
(WTF::setPermissionsOfConfigPage):
(WTF::Config::initialize):

Canonical link: https://commits.webkit.org/292440@main
